File: elixir/test/kyccentral/kyc_test.exs
defmodule KYCCentral.KYCTest do
  @moduledoc "Assessment parsing, argument validation and queued-job polling."

  use ExUnit.Case, async: true

  alias KYCCentral.{Assessment, Error, Stub}

  describe "assess/3" do
    test "parses the response into a typed assessment" do
      {client, _agent} = Stub.client([Stub.json(Stub.assessment_payload())])

      assert {:ok, %Assessment{} = assessment} = KYCCentral.KYC.assess(client, "00445790")

      assert assessment.company_name == "TESCO PLC"
      assert assessment.risk_level == :medium
      assert assessment.psc_chain_depth == 2

      assert Enum.map(assessment.flags, & &1.code) == [
               "ACCOUNTS_OVERDUE",
               "ADVERSE_MEDIA_UNCONFIRMED"
             ]

      # Anything not mapped to a field is still reachable.
      assert assessment.raw["profile"]["company_status"] == "active"
    end

    test "exposes helpers over the flags" do
      {client, _agent} = Stub.client([Stub.json(Stub.assessment_payload())])
      {:ok, assessment} = KYCCentral.KYC.assess(client, "00445790")

      refute Assessment.clear?(assessment)
      refute Assessment.partial?(assessment)
      assert Assessment.flags_at(assessment, [:critical]) == []

      assert Enum.map(Assessment.flags_at_or_above(assessment, :high), & &1.code) ==
               ["ACCOUNTS_OVERDUE"]

      assert Assessment.has_flag?(assessment, "ACCOUNTS_OVERDUE")
      refute Assessment.has_flag?(assessment, "NOT_RAISED")
      assert Assessment.flag(assessment, "NOT_RAISED") == nil
      assert %{code: "ACCOUNTS_OVERDUE"} = Assessment.flag(assessment, "ACCOUNTS_OVERDUE")

      assert Enum.map(Assessment.rules_with_status(assessment, :passed), & &1.code) ==
               ["COMPANY_NOT_ACTIVE"]
    end

    test "marks assessments affected by an upstream timeout as partial" do
      payload = Map.put(Stub.assessment_payload(), "timed_out_services", ["sanctions"])
      {client, _agent} = Stub.client([Stub.json(payload)])

      assert {:ok, assessment} = KYCCentral.KYC.assess(client, "00445790")
      assert Assessment.partial?(assessment)
    end

    test "tolerates a severity this client version predates" do
      payload = Stub.assessment_payload()
      flags = List.update_at(payload["flags"], 0, &Map.put(&1, "severity", "catastrophic"))
      {client, _agent} = Stub.client([Stub.json(Map.put(payload, "flags", flags))])

      assert {:ok, assessment} = KYCCentral.KYC.assess(client, "00445790")
      assert hd(assessment.flags).severity == :low
      assert hd(assessment.flags).raw["severity"] == "catastrophic"
    end

    test "requires exactly one of a company number and :q" do
      {client, agent} = Stub.client([Stub.json(Stub.assessment_payload())])

      assert {:error, %Error{kind: :invalid_argument, message: message}} =
               KYCCentral.KYC.assess(client, [])

      assert message =~ "either"

      assert {:error, %Error{kind: :invalid_argument, message: both}} =
               KYCCentral.KYC.assess(client, "00445790", q: "tesco")

      assert both =~ "not both"
      assert Stub.call_count(agent) == 0
    end

    test "assesses the top search result when given a name" do
      {client, agent} = Stub.client([Stub.json(Stub.assessment_payload())])

      assert {:ok, _} = KYCCentral.KYC.assess(client, q: "tesco plc", rule_set_id: "quick")

      assert Stub.query_value(agent, "q") == "tesco plc"
      assert Stub.query_value(agent, "rule_set_id") == "quick"
      assert Stub.query_value(agent, "company_number") == nil
    end
  end

  describe "queued assessments" do
    defp queued, do: Stub.json(%{"job_id" => "job-123", "status" => "queued"}, 202)

    test "polls a queued job to completion" do
      {client, agent} =
        Stub.client([
          queued(),
          Stub.json(%{"job_id" => "job-123", "status" => "running"}),
          Stub.json(%{
            "job_id" => "job-123",
            "status" => "done",
            "result" => Stub.assessment_payload()
          })
        ])

      assert {:ok, %Assessment{company_name: "TESCO PLC"}} =
               KYCCentral.KYC.assess(client, "00445790", poll_interval: 0)

      assert Stub.call_count(agent) == 3
      assert Stub.path(agent, 1) == "/v1/jobs/job-123"
    end

    test "returns the raw envelope when wait is false" do
      {client, _agent} = Stub.client([queued()])

      assert {:ok, %{"job_id" => "job-123", "status" => "queued"}} =
               KYCCentral.KYC.assess(client, "00445790", wait: false)
    end

    test "reports a failed job" do
      {client, _agent} =
        Stub.client([
          queued(),
          Stub.json(%{"job_id" => "job-123", "status" => "error", "error" => "upstream exploded"})
        ])

      assert {:error, %Error{kind: :job_failed, job_id: "job-123", message: message}} =
               KYCCentral.KYC.assess(client, "00445790", poll_interval: 0)

      assert message =~ "upstream exploded"
    end

    test "reports an expired result" do
      {client, _agent} =
        Stub.client([queued(), Stub.json(%{"job_id" => "job-123", "status" => "done"})])

      assert {:error, %Error{kind: :job_failed, message: message}} =
               KYCCentral.KYC.assess(client, "00445790", poll_interval: 0)

      assert message =~ "no longer available"
    end

    test "gives up after the poll timeout" do
      {client, _agent} =
        Stub.client([queued(), Stub.json(%{"job_id" => "job-123", "status" => "running"})])

      assert {:error, %Error{kind: :job_timeout, job_id: "job-123"}} =
               KYCCentral.KYC.assess(client, "00445790", poll_interval: 1, poll_timeout: 0)
    end
  end
end

File: javascript/test/kyc.test.ts
/** Assessment parsing, argument validation and queued-job polling. */

import { describe, expect, it } from 'vitest';

import {
  JobFailedError,
  JobTimeoutError,
  KYCCentral,
  findFlag,
  flagsAt,
  flagsAtOrAbove,
  hasFlag,
  isClear,
  isPartial,
  type Assessment,
} from '../src/index.js';
import { BASE_URL, assessmentPayload, jsonResponse, mockFetch } from './helpers.js';

function clientWith(responses: Response[]): KYCCentral {
  return new KYCCentral({ apiKey: 'k', baseUrl: BASE_URL, fetch: mockFetch(responses) });
}

describe('assess', () => {
  it('parses the response into a typed assessment', async () => {
    const client = clientWith([jsonResponse(assessmentPayload())]);
    const assessment = await client.kyc.assess('00445790');

    expect(assessment.companyName).toBe('TESCO PLC');
    expect(assessment.riskLevel).toBe('medium');
    expect(assessment.pscChainDepth).toBe(2);
    expect(assessment.flags.map((f) => f.code)).toEqual([
      'ACCOUNTS_OVERDUE',
      'ADVERSE_MEDIA_UNCONFIRMED',
    ]);
    // Anything not mapped to a field is still reachable.
    expect(assessment.raw.profile.company_status).toBe('active');
  });

  it('exposes helpers over the flags', async () => {
    const client = clientWith([jsonResponse(assessmentPayload())]);
    const assessment = await client.kyc.assess('00445790');

    expect(isClear(assessment)).toBe(false);
    expect(isPartial(assessment)).toBe(false);
    expect(flagsAt(assessment, 'critical')).toEqual([]);
    expect(flagsAtOrAbove(assessment, 'high').map((f) => f.code)).toEqual(['ACCOUNTS_OVERDUE']);
    expect(flagsAtOrAbove(assessment, 'medium').map((f) => f.code)).toEqual(['ACCOUNTS_OVERDUE']);
    expect(hasFlag(assessment, 'ACCOUNTS_OVERDUE')).toBe(true);
    expect(hasFlag(assessment, 'NOT_RAISED')).toBe(false);
    expect(findFlag(assessment, 'NOT_RAISED')).toBeUndefined();
  });

  it('marks assessments affected by an upstream timeout as partial', async () => {
    const payload = assessmentPayload();
    payload.timed_out_services = ['sanctions'];
    const client = clientWith([jsonResponse(payload)]);

    expect(isPartial(await client.kyc.assess('00445790'))).toBe(true);
  });

  it('tolerates a severity this client version predates', async () => {
    const payload = assessmentPayload();
    payload.flags[0].severity = 'catastrophic';
    const client = clientWith([jsonResponse(payload)]);

    const assessment = await client.kyc.assess('00445790');
    expect(assessment.flags[0]!.severity).toBe('low');
    expect(assessment.raw.flags[0].severity).toBe('catastrophic');
  });

  it('requires exactly one of a company number and q', async () => {
    const client = clientWith([jsonResponse(assessmentPayload())]);

    await expect(client.kyc.assess({ q: '' } as never)).rejects.toThrow(/either/);
    await expect(client.kyc.assess('00445790', { q: 'tesco' })).rejects.toThrow(/not both/);
  });

  it('assesses the top search result when given a name', async () => {
    const fetch = mockFetch([jsonResponse(assessmentPayload())]);
    const client = new KYCCentral({ apiKey: 'k', baseUrl: BASE_URL, fetch });

    await client.kyc.assess({ q: 'tesco plc', ruleSetId: 'quick' });

    const url = new URL(fetch.mock.calls[0]![0]);
    expect(url.searchParams.get('q')).toBe('tesco plc');
    expect(url.searchParams.get('rule_set_id')).toBe('quick');
    expect(url.searchParams.has('company_number')).toBe(false);
  });
});

describe('queued assessments', () => {
  const queued = () => jsonResponse({ job_id: 'job-123', status: 'queued' }, 202);

  it('polls a queued job to completion', async () => {
    const fetch = mockFetch([
      queued(),
      jsonResponse({ job_id: 'job-123', status: 'running' }),
      jsonResponse({ job_id: 'job-123', status: 'done', result: assessmentPayload() }),
    ]);
    const client = new KYCCentral({ apiKey: 'k', baseUrl: BASE_URL, fetch });

    const assessment = (await client.kyc.assess('00445790', {
      pollIntervalMs: 0,
    })) as Assessment;

    expect(assessment.companyName).toBe('TESCO PLC');
    expect(fetch).toHaveBeenCalledTimes(3);
    expect(fetch.mock.calls[1]![0]).toBe(`${BASE_URL}/v1/jobs/job-123`);
  });

  it('returns the raw envelope when wait is false', async () => {
    const client = clientWith([queued()]);
    const result = await client.kyc.assess('00445790', { wait: false });

    expect(result).toEqual({ job_id: 'job-123', status: 'queued' });
  });

  it('throws when the job fails', async () => {
    const client = clientWith([
      queued(),
      jsonResponse({ job_id: 'job-123', status: 'error', error: 'upstream exploded' }),
    ]);

    await expect(client.kyc.assess('00445790', { pollIntervalMs: 0 })).rejects.toThrow(
      JobFailedError,
    );
  });

  it('throws when the result has expired', async () => {
    const client = clientWith([queued(), jsonResponse({ job_id: 'job-123', status: 'done' })]);

    await expect(client.kyc.assess('00445790', { pollIntervalMs: 0 })).rejects.toThrow(
      /no longer available/,
    );
  });

  it('gives up after the poll timeout', async () => {
    const client = clientWith([queued(), jsonResponse({ job_id: 'job-123', status: 'running' })]);

    await expect(
      client.kyc.assess('00445790', { pollIntervalMs: 1, pollTimeoutMs: 0 }),
    ).rejects.toThrow(JobTimeoutError);
  });
});
